notebooks/python/1.3. BigQuery Storage & Spark ML.ipynb

```diff
@@ -0,0 +1,385 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "cat: /release: No such file or directory\n",
+      "Scala code runner version 2.11.12 -- Copyright 2002-2017, LAMP/EPFL\n"
+     ]
+    }
+   ],
+   "source": [
+    "!scala -version"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create Spark Session\n",
+    "\n",
+    "Include the correct version of the spark-bigquery-connector jar\n",
+    "\n",
+    "If you are using scala version 2.11 use `'gs://spark-lib/bigquery/spark-bigquery-latest.jar'`.\n",
+    "\n",
+    "If you are using scala version 2.12 use `'gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar'`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "'2.4.5'"
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from pyspark.sql import SparkSession\n",
+    "\n",
+    "spark = SparkSession.builder \\\n",
+    "  .appName('BigQuery Storage &  Spark ML')\\\n",
+    "  .config('spark.jars', 'gs://spark-lib/bigquery/spark-bigquery-latest.jar') \\\n",
+    "  .getOrCreate()\n",
+    "\n",
+    "spark.version"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Read the data from BigQuery as a Spark Dataframe.\n",
+    "table  = 'bigquery-public-data.samples.natality'\n",
+    "\n",
+    "df_natality_table = spark.read \\\n",
+    "  .format(\"bigquery\") \\\n",
+    "  .option(\"table\", table) \\\n",
+    "  .load()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "root\n",
+      " |-- weight_pounds: double (nullable = true)\n",
+      " |-- mother_age: long (nullable = true)\n",
+      " |-- father_age: long (nullable = true)\n",
+      " |-- gestation_weeks: long (nullable = true)\n",
+      " |-- weight_gain_pounds: long (nullable = true)\n",
+      " |-- apgar_5min: long (nullable = true)\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "# limit no of rows that will be read for demo \n",
+    "limit = 1000\n",
+    "\n",
+    "df_natality_select = df_natality_table \\\n",
+    ".select(\"weight_pounds\", \"mother_age\", \"father_age\", \"gestation_weeks\", \"weight_gain_pounds\", \"apgar_5min\") \\\n",
+    ".where(\"\"\"\n",
+    "weight_pounds IS NOT NULL \n",
+    "AND mother_age IS NOT NULL\n",
+    "AND father_age IS NOT NULL\n",
+    "AND gestation_weeks IS NOT NULL\n",
+    "AND weight_gain_pounds IS NOT NULL\n",
+    "AND apgar_5min IS NOT NULL\n",
+    "\"\"\") \\\n",
+    ".limit(limit) \\\n",
+    ".cache()\n",
```

Comment:
It's worth adding a comment to this line to explain why `.cache()` is necessary (LinearRegression is iterative and you want to avoid re-reading the data from GCS repeatedly)